rslib/src/card_rendering/parser.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::collections::HashMap;
+
+use nom::{
+    branch::alt,
+    bytes::complete::is_not,
+    bytes::complete::tag,
+    character::complete::{anychar, multispace0},
+    combinator::{map, not, recognize, success, value},
+    multi::{many0, many1},
+    sequence::{delimited, pair, preceded, separated_pair, terminated, tuple},
+};
+
+use super::{CardNodes, Node, OtherTag, Tag, TtsTag};
+
+type IResult<'a, O> = nom::IResult<&'a str, O>;
+
+impl<'a> CardNodes<'a> {
+    pub(super) fn parse(mut txt: &'a str) -> Self {
+        let mut nodes = Vec::new();
+        while let Ok((remaining, node)) = node(txt) {
+            txt = remaining;
+            nodes.push(node);
+        }
+
+        Self(nodes)
+    }
+}
+
+impl<'a> Tag<'a> {
+    fn new(name: &'a str, options: Vec<(&'a str, &'a str)>, content: &'a str) -> Self {
+        match name {
+            "tts" => {
+                let mut lang = "";
+                let mut voices = vec![];
+                let mut speed = 1.0;
+                let mut blank = None;
+                let mut other_options = HashMap::new();
+
+                for option in options {
+                    match option.0 {
+                        "lang" => lang = option.1,
+                        "voices" => voices = option.1.split(',').collect(),
+                        "speed" => speed = option.1.parse().unwrap_or(1.0),
+                        "cloze_blank" => blank = Some(option.1),
+                        _ => {
+                            other_options.insert(option.0, option.1);
+                        }
+                    }
+                }
+
+                Self::Tts(TtsTag {
+                    content,
+                    lang,
+                    voices,
+                    speed,
+                    blank,
+                    options: other_options,
+                })
+            }
+            _ => Self::Other(OtherTag {
+                name,
+                content,
+                options: options.into_iter().collect(),
+            }),
+        }
+    }
+}
+
+/// Consume 0 or more of anything in " \t\r\n" after `parser`.
+fn trailing_whitespace0<'parser, 's, P, O>(parser: P) -> impl FnMut(&'s str) -> IResult<O>
+where
+    P: FnMut(&'s str) -> IResult<O> + 'parser,
+{
+    terminated(parser, multispace0)
+}
+
+/// Parse until char in `arr` is found. Always succeeds.
+fn is_not0<'parser, 'arr: 'parser, 's: 'parser>(
+    arr: &'arr str,
+) -> impl FnMut(&'s str) -> IResult<&'s str> + 'parser {
+    alt((is_not(arr), success("")))
+}
+
+fn node(s: &str) -> IResult<Node> {
+    alt((text_node, sound_node, tag_node))(s)
+}
+
+/// A sound tag `[sound:ressource]`, where `ressource` is pointing to a sound or video file.
```

Comment:
Very minor: there are few instances that should read 'resource' in this and another file